packages/astro/src/core/render/dev/css.ts

```diff
@@ -13,41 +13,50 @@ const cssRe = new RegExp(
 );
 export const isCSSRequest = (request: string): boolean => cssRe.test(request);
 
-/**
- * getStylesForURL
- * Given a filePath URL, crawl Vite’s module graph to find style files
- */
+/** Given a filePath URL, crawl Vite’s module graph to find all style imports. */
 export function getStylesForURL(filePath: URL, viteServer: vite.ViteDevServer): Set<string> {
-	const css = new Set<string>();
+	const importedCssUrls = new Set<string>();
 
-	// recursively crawl module graph to get all style files imported by parent id
-	function crawlCSS(id: string, scanned = new Set<string>()) {
-		// note: use .getModulesByFile() to get all related nodes of the same URL
-		// using .getModuleById() could cause missing style imports on initial server load
-		const relatedMods = viteServer.moduleGraph.getModulesByFile(id) ?? new Set();
+	/** recursively crawl the module graph to get all style files imported by parent id */
+	function crawlCSS(_id: string, isFile: boolean, scanned = new Set<string>()) {
+		const id = unwrapId(_id);
 		const importedModules = new Set<vite.ModuleNode>();
-
-		for (const relatedMod of relatedMods) {
-			if (id === relatedMod.id) {
+		const moduleEntriesForId = isFile
+			? // If isFile = true, then you are at the root of your module import tree.
+			  // The `id` arg is a filepath, so use `getModulesByFile()` to collect all
+			  // nodes for that file. This is needed for advanced imports like Tailwind.
+			  viteServer.moduleGraph.getModulesByFile(id) ?? new Set()
+			: // Otherwise, you are following an import in the module import tree.
+			  // You are safe to use getModuleById() here because Vite has already
+			  // resolved the correct `id` for you, by creating the import you followed here.
+			  new Set([viteServer.moduleGraph.getModuleById(id)!]);
```

Comment:
Huh, nice find! Glad we can default to the ID-based system where possible

Reply:
Doh, I meant to call this out to you! The reason for the change was that we needed to support virtual imports that contain CSS but don't exist on file, like the ones added by our Tailwind hooks.

I'd love your help confirming this change is good, it made sense to me and I tested it more than any other part of this PR probably.